src/gpuarray_buffer_collectives.c

```diff
@@ -4,43 +4,58 @@
 #include "gpuarray/buffer_collectives.h"
 #include "gpuarray/error.h"
 
-int gpucomm_new(gpucomm* comm, gpucontext* ctx, const char* clique_id,
-                     int ndev, int rank, int* res) {
-  comm = NULL;
-  if (ctx->comm_ops == NULL)
+int gpucomm_new(gpucomm** comm, gpucontext* ctx,
+                gpucommCliqueId cliqueId, int ndev, int rank) {
+  if (ctx->comm_ops == NULL) {
+    *comm = NULL;
     return GA_UNSUPPORTED_ERROR;
-  return ctx->comm_ops->comm_new(comm, ctx, clique_id, ndev, rank, res);
+  }
+  return ctx->comm_ops->comm_new(comm, ctx, cliqueId, ndev, rank);
 }
 
-void gpucomm_free(gpucomm* comm) {
-  gpucomm_context(comm)->comm_ops->comm_free(comm);
+int gpucomm_free(gpucomm* comm) {
+  gpucontext* ctx = gpucomm_context(comm);
+  if (ctx->comm_ops == NULL)
+    return GA_COMM_ERROR;
+  ctx->comm_ops->comm_free(comm);
+  return GA_NO_ERROR;
 }
 
-const char* gpucomm_error(gpucomm* comm, int err) {
-  if (err == GA_COMM_ERROR && comm != NULL)
-    return gpucomm_context(comm)->comm_ops->comm_error(gpucomm_context(comm));
-  else
-    return gpucontext_error(gpucomm_context(comm), err);
+const char* gpucomm_error(gpucontext* ctx) {
+  if (ctx->comm_ops != NULL)
+      return ctx->comm_ops->comm_error(ctx);
+  return "No collective ops available, API error. Is a collectives library installed?";
 }
 
 gpucontext* gpucomm_context(gpucomm* comm) {
   return ((partial_gpucomm*)comm)->ctx;
 }
 
-const char* gpucomm_gen_clique_id(int* res) {
-  return gpucomm_context(comm)->comm_ops->generate_clique_id(res);
+int gpucomm_gen_clique_id(gpucontext* ctx, gpucommCliqueId* cliqueId) {
+  if (ctx->comm_ops == NULL)
+    return GA_COMM_ERROR;
+  return ctx->comm_ops->generate_clique_id(ctx, cliqueId);
 }
 
-int gpucomm_get_count(const gpucomm* comm, int* count) {
-  return gpucomm_context(comm)->comm_ops->get_count(comm, count);
+int gpucomm_get_count(gpucomm* comm, int* count) {
+  gpucontext* ctx = gpucomm_context(comm);
+  if (ctx->comm_ops == NULL)
+    return GA_COMM_ERROR;
+  return ctx->comm_ops->get_count(comm, count);
 }
 
-int gpucomm_get_device(const gpucomm* comm, int* device) {
-  return gpucomm_context(comm)->comm_ops->get_device(comm, device);
+int gpucomm_get_device(gpucomm* comm, int* device) {
+  gpucontext* ctx = gpucomm_context(comm);
+  if (ctx->comm_ops == NULL)
+    return GA_COMM_ERROR;
+  return ctx->comm_ops->get_device(comm, device);
```

Comment:
Actually, since you have gpucomm_context, it should probably just not be present.